Add CI workflow

lourencofsilva opened the pull request:
Add workflow that builds, run tests and lint on the repo. This is initiated by a push or pull request to the develop branch. Resolves #11

lourencofsilva committed:
Add CI workflow

Add workflow that builds, run tests and lint on the repo. This is initiated by a push or pull request to the develop branch.

suzanneEmbury commented on .github/workflows/build-lint-test.yml:
```diff
@@ -0,0 +1,41 @@
+
+name: Build, Lint, and Test
+
+on:
```
Looks good.  Can we also have the workflow_dispatch trigger?  It's often useful to be able to manually trigger a workflow, especially in the early days when we are figuring out how to set it up.

lourencofsilva replied:
I completely agree. I will add the workflow_dispatch trigger, and related to the discussion below on Python version, will add an option to select a specific Python version on this manual trigger so that specific versions can be tested if desired.

suzanneEmbury replied:
Great - thanks!

suzanneEmbury commented on .github/workflows/build-lint-test.yml:
```diff
@@ -0,0 +1,41 @@
+    - name: Set up Python 3.11
+      uses: actions/setup-python@v3
+      with:
+        python-version: "3.11"
```
Good idea to set this to match our Jenkins set up on our mirror repo on GitLab.

Is there a way to have the GitHub Actions workflow pick up the Python version set in the project config?  This may be a pipe dream, but it feels like it could be useful to have the CI pipeline run on whatever we set as the lowest compatibility of Python in the project config.

No problem if you don't have time to look into this.  I can pick it up.

lourencofsilva replied:
No worries, I have implemented this. The version flow is as follows:

1. The Python version can be specified using the workflow_dispatch input python-version, on manual runs.
2. If not specified, it will be read from pyproject.toml, using the lowest compatibility.
3. In case of no specification found, will default to 3.11.

suzanneEmbury commented on .github/workflows/build-lint-test.yml:
```diff
@@ -0,0 +1,41 @@
+    steps:
+    - uses: actions/checkout@v4
+    - name: Set up Python 3.11
+      uses: actions/setup-python@v3
```
Is there a reason you picked v3 for this action?  I think the action itself is currently up to v5 (https://github.com/actions/setup-python) but I didn't find the release changes list to tell me whether needed to be on v5 or not.  So fine if v3 is the right choice.

lourencofsilva replied:
There was no reason at all, it was a consequence of my following of the GitHub actions template (which doesn't seem to have been updated). It is in fact on v5 and I should be using that. Testing showed no issue with v5 so will update.

suzanneEmbury replied:
Great - thanks!

suzanneEmbury commented on .github/workflows/build-lint-test.yml:
```diff
@@ -0,0 +1,41 @@
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install flake8 pytest
```
Happy to go with Flake8 for now - though I've been wondering about trying out Ruff.  Have you had any experience with it?

lourencofsilva replied:
I agree with the suggestion. Ruff seems like a good updated option, joined with its extremely fast run time (perks of anything written in Rust). I have updated the action to use Ruff. At the moment, it is using its default configuration and treating all errors as warnings, so that the action completes and simply displays them.

There are several improvements that can be implemented, however these are at personal discretion so I've left them here as options:

1. The specific flakes, errors and warnings being check for, as well as all other Ruff config can be manually set in pyproject.toml and it will pick them up automatically.
2. The action of ruff check can be split into two, where the first one selects only a few specific rules to be checked for and make the action error if these are found (for more severe errors such as syntax errors), and then having a second check for all other desired rules that only serves as warnings (as the current setup does).
3. A few quick lines of code can be added for Ruff to automatically fix the errors found (those it can fix automatically) and commit these either to the codebase or the pull request, depending on the situation.

suzanneEmbury replied:
Thanks for the suggestions.  Maybe you could make an issue to capture this into the future?  Or I can copy your text (making it clear you are the author) into an issue if you prefer.  I'd like to live with Ruff for a bit in its most basic form, before trying anything more sophisticated with it.

lourencofsilva replied:
Of course, I'll capture this in a new issue. Issue #18

suzanneEmbury commented on .github/workflows/build-lint-test.yml:
```diff
@@ -0,0 +1,41 @@
+        pip install flake8 pytest
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+    - name: Set PYTHONPATH
+      run: echo "PYTHONPATH=${{ github.workspace }}" >> $GITHUB_ENV
```
Do we need to set the PYTHONPATH here?  Which bit of the workflow fails if we don't have it?

lourencofsilva replied:
We do need PYTHONPATH setup, or else pytest fails as it cannot import robota-core packages while running. I added this as a fix during prior testing.

nigelmj replied:
I had the same problem when working on this issue (I didn't create a PR as this was created already). My fix was using `python -m pytest` rather than calling `pytest`. This meant PYTHONPATH was no longer required as it ensures the tests are in the current python enviroment. Maybe this could be used instead?

suzanneEmbury replied:
Thanks very much @Nigel007 - one of the delays in my next review of this PR was that I was trying to figure out why none of our other Python GA workflows needed the PYTHONPATH like this.

I think I'd prefer to go with the solution suggested by @Nigel007 , even though @lourencofsilva 's solution is also plausible, so that the GA workflow set up matches the dev environment set up as closely as possible (where we don't set PYTHONPATH explicitly either).

lourencofsilva replied:
I agree with this, `python -m pytest` is the best way of going about this. Thanks for the input @Nigel007. I have just altered this and removed the PYTHONPATH setup.

suzanneEmbury reviewed (changes requested):
This is great - thanks so much for taking the time to propose this.  I have a few queries and then hopefully this will be ready to merge.

lourencofsilva committed:
Alter build-lint-test action

Add workflow_dispatch manual trigger, with option of specifying a python version. Default to minimum required version in pyproject.toml. Update setup-python to v5. Testing lint with ruff.

nigelmj commented:
Have you considered adding mypy for type checking? It could help catch potential type-related issues early on, especially since the codebase already utilises the typing module. Just a thought!

suzanneEmbury commented:
> Have you considered adding mypy for type checking? It could help catch potential type-related issues early on, especially since the codebase already utilises the typing module. Just a thought!

This is a good suggestion, @Nigel007 - we have used mypy in other projects and found it useful.  Maybe you could add this in a separate issue/PR?  I'd like to merge this PR as soon as possible, since @lourencofsilva created it some time ago and I've already delayed it for a while.

suzanneEmbury reviewed (changes requested):
This is getting really close now, @lourencofsilva .  Thanks so much for your great work on this.

lourencofsilva committed:
Remove PYTHONPATH

Altered GA workflow so that PYTHONPATH is not setup, and run tests through `python -m` instead of directly calling `pytest`

suzanneEmbury reviewed (approved):
I think I'm happy with this now, @lourencofsilva - thanks.  It took me a while to convince myself that the ${{...}} were safe but I think they are, provided I keep an eye on PRs that change the pyproject.toml file (which I'd have to do anyway).

Thanks again for your hard work on this.
